fmn/cache/rules.py

```diff
@@ -0,0 +1,52 @@
+import logging
+from typing import TYPE_CHECKING
+
+from cashews import cache
+from cashews.formatter import get_templates_for_func
+
+from ..database.model import Rule
+from .util import cache_early_ttl, cache_ttl
+
+if TYPE_CHECKING:
+    from fedora_messaging.message import Message
+    from sqlalchemy.ext.asyncio import AsyncSession
+
+log = logging.getLogger(__name__)
+
+
+class RulesCache:
+    """Cache the rules currently in the database."""
+
+    def __init__(self):
+        self._db: "AsyncSession" | None = None
+
+    def set_db(self, db: "AsyncSession"):
+        self._db = db
+
+    @cache.locked(key="rules", prefix="v1", ttl="1h")
+    async def build(self):
+        if self._db is None:
+            raise RuntimeError("You must use set_db() first.")
+        log.debug("Building the rules cache")
+        result = await self._db.execute(Rule.select_related().filter_by(disabled=False))
+        log.debug("Built the rules cache")
+        return list(result.scalars())
+
+    @cache.early(
+        key="rules", prefix="v1", ttl=cache_ttl("rules"), early_ttl=cache_early_ttl("rules")
+    )
+    async def get_rules(self):
+        return await self.build()
+
+    async def invalidate(self):
+        log.debug("Invalidating the rules cache")
+        cache_key = list(get_templates_for_func(self.get_rules))[0]
+        await cache.delete(cache_key)
```

Comment:
I’m not sure why there can be more than one template for a function, but looking at `get_templates_for_func()`, it seems possible and we could cope with that like this:

```
cache_keys = get_templates_for_func(self.get_rules)
await asyncio.gather(*(cache.delete(key) for key in cache_keys))
```

Reply:
Yeah I don't see why we would have more than one template either. But indeed your solution is better.